Change `UnpackError` with `UnpackDefect`

bobeff opened the pull request:
The deprecation warning for `UnpackError` exception raised by some `inline` procedures in the Nim standard library propagates to the user code. If the user code has a requirement for building without warnings this is a problem for the successful execution of the tests. In order to resolve this, all occurrences of `UnpackError` in the Nim code base are changed to `UnpackDefect`. Only the type alias is retained to not break other people's user code since `UnpackError` is exported type.

bobeff committed:
Change `UnpackError` with `UnpackDefect`

The deprecation warning for `UnpackError` exception raised by some
`inline` procedures in the Nim standard library propagates to the user
code. If the user code has a requirement for building without warnings
this is a problem for the successful execution of the tests. In order
to resolve this, all occurrences of `UnpackError` in the Nim code base
are changed to `UnpackDefect`. Only the type alias is retained to not
break other people's user code since `UnpackError` is exported type.

ghost commented:
There might be a problem with backwards compatibility though

bobeff commented:
> There might be a problem with backwards compatibility though

I can't imagine why. I retained the type alias and both `UnpackError` and `UnpackDefect` should be valid in the user code. Can you give an example?

Araq commented:
Your patch is very good but the code was wrong to begin with, one cannot catch an `UnpackDefect`.

Yardanico commented on lib/pure/nativesockets.nim:
```diff
@@ -447,7 +447,7 @@ proc getSockDomain*(socket: SocketHandle): Domain =
   try:
     result = toKnownDomain(name.sin6_family.cint).get()
-  except UnpackError:
+  except UnpackDefect:
```
As @Araq said, I guess you should change this code to something like:
```nim
let res = toKnownDomain(name.sin6_family.cint)
if res.isSome():
  result = res.get()
else:
  raise newException(IOError, "Unknown socket family in getSockDomain")
```

bobeff replied:
It's fixed.

bobeff committed:
Remove the catching of `UnpackDefect`

Defect exceptions should not be cached, because they indicate problem in
the API usage. The code in `nimblesocket.nim` is rewritten to first
check whether there is a value set into the `knownDomain` variable from
the `Option` type before usage.

Yardanico commented on lib/pure/nativesockets.nim:
```diff
@@ -445,9 +445,10 @@ proc getSockDomain*(socket: SocketHandle): Domain =
-  try:
+  let knownDomain = toKnownDomain(name.sin6_family.cint)
+  if knownDomain.isSome:
     result = toKnownDomain(name.sin6_family.cint).get()
```
uhh, this is not correct since you call it again, you should have "knownDomain.get()" instead :P

bobeff replied:
Yes. I'm very careless. :(

bobeff replied:
It's fixed.
